blocks/bricks/__init__.py

```diff
@@ -1,22 +1,324 @@
+    When a :class:`Brick` is instantiated and its application method (i.e.
+    an instance of this class) requested, the descriptor protocol (through
+    the :meth:`__get__` method) automatically instantiates a
+    :class:`BoundApplication` class and returns this. This bound
+    application class can be used to store application information
+    particular to a brick instance. Any attributes unknown to the bounded
+    application are automatically routed to the application that
+    instantiated it.
+
+    """
+    call_stack = []
+
+    def __init__(self, application):
+        self.application = application
+        self.delegate_function = None
+        self.properties = {}
+        self.instances = {}
+
+    def property(self, name):
+        """Decorator to make application properties.
+
+        Parameters
+        ----------
+        name : str
+            The name the property should take.
+
+        Examples
+        --------
+        >>> class Foo(Brick):
+        ...     @application
+        ...     def apply(self, x):
+        ...         return x + 1
+        ...
+        ...     @apply.property('inputs')
+        ...     def apply_inputs(self):
+        ...         return ['foo', 'bar']
+        >>> foo = Foo()
+        >>> foo.apply.inputs
+        ['foo', 'bar']
+
+        """
+        if not isinstance(name, six.string_types):
+            raise ValueError
+
+        def wrap_property(property_):
+            self.properties[name] = property_
+            return property_
+        return wrap_property
+
+    def delegate(self, f):
+        """Decorator to assign a delegate application.
+
+        An application method can assign a delegate application. Whenever
+        an attribute is not available, it will be requested from the
+        delegate instead.
+
+        Examples
+        --------
+        >>> class Foo(Brick):
+        ...     @application(outputs=['baz'])
+        ...     def apply(self, x):
+        ...         return x + 1
+        ...
+        ...     @apply.property('inputs')
+        ...     def apply_inputs(self):
+        ...         return ['foo', 'bar']
+        >>> class Bar(Brick):
+        ...     def __init__(self, foo):
+        ...         self.foo = foo
+        ...
+        ...     @application(outputs=['foo'])
+        ...     def apply(self, x):
+        ...         return x + 1
+        ...
+        ...     @apply.delegate
+        ...     def apply_delegate(self):
+        ...         return self.foo.apply
+        >>> foo = Foo()
+        >>> bar = Bar(foo)
+        >>> bar.apply.outputs
+        ['foo']
+        >>> bar.apply.inputs
+        ['foo', 'bar']
+
+        """
+        self.delegate_function = f
+        return f
+
+    def __get__(self, instance, owner):
+        """Instantiate :class:`BoundedApplication` for each :class:`Brick`."""
+        if instance is None:
+            return self
+        elif instance in self.instances:
+            return self.instances[instance]
```

Comment:
`self.instances[instance]` is a bit confusing. The dictionary should be renamed, I think, `self.bound_applications` perhaps.